### PR TITLE
OOP exercise - done

### PR DESCRIPTION
**Pull request by evgenii-del**
_Complete the form below and remove this line._

### OOP exercise

[Demo](http://changeable-news.surge.sh/) |
[Code base](https://github.com/evgenii-del/a-tiny-JS-world/tree/master)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/evgenii-del/oop-exercise/index.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        this.saying = saying;
+    }
+
+    toString(object) {
+        return [this.species, this.name, this.gender, this.saying].join(';');
+    }
+}
+
+class Man extends Inhabitant {
+    constructor(species, name, gender, saying, hands, legs) {
```
Comment (OleksiyRudenko) [inline submissions/evgenii-del/oop-exercise/index.js:25]: As we are in a `Man` class we already know the species (that we can send as an argument/string to the `super`) and can skip asking a developer to pass the species when calling `new Man`.
Same is applicable to other classes definitions.

File: submissions/evgenii-del/oop-exercise/index.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        this.legs = legs;
+    }
+
+    toString(object) {
+        return super.toString() + [this.hands, this.legs].join(';');
+    }
+}
+
+class Woman extends Inhabitant {
+    constructor(species, name, gender, saying, hands, legs) {
```
Comment (OleksiyRudenko) [inline submissions/evgenii-del/oop-exercise/index.js:37]: `constructor` is a special function, yet a function, which means we can define default values for some of the params that are least unique.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@evgenii-del Good job!
A couple of improvements will make it great.

File: submissions/evgenii-del/oop-exercise/index.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    toString(object) {
+        return [this.species, this.name, this.gender, this.saying].join(';');
+    }
+}
+
+class Man extends Inhabitant {
+    constructor(name, gender, saying = "Hi!", hands = 2, legs = 2) {
+        super('man', name, gender,);
```
Comment (OleksiyRudenko) [inline submissions/evgenii-del/oop-exercise/index.js:26]: Do we really need to specify gender here?
Besides 'man' is not a species. 'Human' is a species.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@evgenii-del 
There was a catch.
The task requested to create a man, a woman, a dog, a cat. The task requested to output species. Well, a man and a woman are not different species )

File: submissions/evgenii-del/oop-exercise/index.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+class Inhabitant {
+    constructor(species, name, gender, saying) {
+        this.species = species;
+        this.name = name;
+        this.gender = gender;
+        this.saying = saying;
+    }
+
+    toString(object) {
```
Comment (OleksiyRudenko) [inline submissions/evgenii-del/oop-exercise/index.js:19]: Unused variable `object`?

File: submissions/evgenii-del/oop-exercise/index.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        return [this.species, this.name, this.gender, this.saying].join(';');
+    }
+}
+
+class Man extends Inhabitant {
+    constructor(name, gender, saying = "Hi!", hands = 2, legs = 2) {
```
Comment (OleksiyRudenko) [inline submissions/evgenii-del/oop-exercise/index.js:25]: Doesn't `Man` imply gender? Yet unused variable.
Please check your code for code optimizations when refactoring.
Here we've got quite small amount of code. Tech debt may affect code scalability.

File: submissions/evgenii-del/oop-exercise/index.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+const inhabitants = [
+    new Man('John'),
+    new Woman('Emma'),
+    new Cat('Sam', 'male'),
+    new Dog('Bob', 'female')
+];
+
+inhabitants.forEach(obj => print(obj));
```
Comment (OleksiyRudenko) [inline submissions/evgenii-del/oop-exercise/index.js:77]: Bonus learning for you: try `.forEach(print)`. `print` definition already fits required callback specification

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@evgenii-del well done!
You use quite advanced language features like `super.methodName`. You have quite good understanding of hierarchy design, which is a good skill to have to design systems.

Let's polish the code so that future reviewers (e.g. from outside this course) wouldn't have an opportunity to pick those tiny things.

File: submissions/evgenii-del/oop-exercise/index.js (hunk 6 above)
Comment (evgenii-del) [reply submissions/evgenii-del/oop-exercise/index.js]: It doesn't work for me. I may have misunderstood

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
The output looks not sufficiently structured.
Looks like a saying contains a number.

image

Can you please fix this tiny glitch?

**Review by OleksiyRudenko — APPROVED** [review]
Well done!

P.S. Please keep your demos up to date with code change.

**Pull request merged**